### PR TITLE
feat: checkins-sync command to synchronize checkins from config file

### PR DESCRIPTION
**Pull request by subzero10**
## Status
**READY**

## Description
Adds a new command `honeybadger:checkins:sync` to synchronize checkins from configuration file with Honeybadger's Checkins API.

## Related PRs
- The PHP [PR](https://github.com/honeybadger-io/honeybadger-php/pull/173)

## Todos
- [x] Tests
- [x] Update honeybadger-php version to stable version after this [PR](https://github.com/honeybadger-io/honeybadger-php/pull/173) is merged and released
- [x] Changelog Entry (unreleased)
- [x] Create artisan command
- [x] Add configuration entries (`personal_auth_token`, `checkins`)

### CODE REVIEW

**Comment by subzero10** [comment]
Progress!
I would appreciate any feedback around:
- code conventions (I haven't coded in PHP for a while)
- the output of the `honeybadger:checkins-sync` command

Screenshot 2023-09-30 at 12 19 44 AM

Screenshot 2023-09-30 at 12 19 58 AM
2/c7377156-372c-46ac-9b2e-d0e37e57f847">

File: composer.json
Comment (BethanyBerkowitz) [inline composer.json:25]: Did you mean to check this in?

File: config/honeybadger.php
Comment (BethanyBerkowitz) [inline config/honeybadger.php:138]: This is helpful, you may want to include it in the README too. 

I looked at the docs you linked to and they could use a bit more information if we're allowing folks to set up their checkins like this instead of via the UI. Is there an issue for that?

**Review by BethanyBerkowitz — COMMENTED** [review]
I can't comment on PHP code conventions, will defer to @shalvah. But I love the output table, it's excellent.

File: src/HoneybadgerServiceProvider.php
Comment (shalvah) [inline src/HoneybadgerServiceProvider.php:192]: I don't think there's a need for this error handling. The point of the service exception handler is to prevent runtime errors from disrupting our users. Things like this check in task are one-time tasks: immediate error feedback should be provided in the console.

File: src/Commands/HoneybadgerCheckinsSyncCommand.php
Comment (shalvah) [inline src/Commands/HoneybadgerCheckinsSyncCommand.php:49]: Tbh, I don't think this is needed. My recommendation: if there are specific errors that we can anticipate (if any), then handle only those—print a useful message + help, and let all others crash.

The reason is that by printing only the exception message here, you're effectively hiding information that might be helpful to the user (the stack trace). That's worse than just letting it crash (in which case it will print a stack trace).

File: src/HoneybadgerServiceProvider.php
Comment (shalvah) [inline src/HoneybadgerServiceProvider.php:101]: I'm not 100% certain, but I don't think this is needed. I think you can just pass the list of command classes to `$this->commands`.

File: composer.json
Comment (shalvah) [reply composer.json:25]: Yeah, that's one thing I dislike about the composer local surrogacy—you have to remember to not check it in. Also the repositories entry below.

**Review by shalvah — APPROVED** [review]
Hehe, this is pretty good! A few things I think may be unnecessary. Plus, a word on naming: I find CheckinsSync a bit weird. I think `SyncCheckins` is better name (command = action = verb-noun). I'd also recommend making the command string `checkins:sync`.

File: composer.json
Comment (subzero10) [reply composer.json:25]: Before merging, I will set the correct dependency version, but it has to be after we merge and release the honeybadger-php package.

File: config/honeybadger.php
Comment (subzero10) [reply config/honeybadger.php:138]: Yes, there an [issue](https://github.com/honeybadger-io/docs/issues/346) to update PHP/Laravel documentation. I will work on that asap!

**Comment by subzero10** [comment]
> Hehe, this is pretty good! A few things I think may be unnecessary. Plus, a word on naming: I find CheckinsSync a bit weird. I think `SyncCheckins` is better name (command = action = verb-noun). I'd also recommend making the command string `checkins:sync`.

I struggled to find a good name, I will use `SyncCheckins`!
In regards to the command, we already have a command `honeybadger:checkin` and I think it's important to have the `honeybadger` prefix. Maybe `honeybadger:sync-checkins` ?

File: src/Commands/HoneybadgerCheckinsSyncCommand.php
Comment (subzero10) [reply src/Commands/HoneybadgerCheckinsSyncCommand.php:49]: > Tbh, I don't think this is needed. My recommendation: if there are specific errors that we can anticipate (if any), then handle only those—print a useful message + help, and let all others crash.

I wasn't sure how to tackle this and I went for the approach that was already implemented for the rest of the package: let the user control whether an exception in the honeybadger will cause the script to crash or simply log a message.
If there is a `ServiceException` is thrown, `$e->getMessage()` provides a helpful message to the user.

> The reason is that by printing only the exception message here, you're effectively hiding information that might be helpful to the user (the stack trace). That's worse than just letting it crash (in which case it will print a stack trace).

Good point, will correct!

In conclusion, here's what I will do:
- I will catch `ServiceException` and log a warning message
- I will let other exceptions crash the process and (potentially) blocking the deployment script.

Comment (subzero10) [reply src/Commands/HoneybadgerCheckinsSyncCommand.php:49]: > I will let other exceptions crash the process and (potentially) blocking the deployment script.

Actually, I'm not sure about this. If there are other kind of exceptions thrown, it probably means that they are out of our control (and quite possible the user's control). In that case, why block the deployment script? What do you think?

File: src/HoneybadgerServiceProvider.php
Comment (subzero10) [reply src/HoneybadgerServiceProvider.php:101]: I just followed the way other commands are defined here 🤷‍♂️

File: src/HoneybadgerServiceProvider.php
Comment (subzero10) [reply src/HoneybadgerServiceProvider.php:192]: We're not hiding the feedback in the console - the message is printed in the console.
My approach was to let the user decide whether an error within Honeybadger would stop their deployment script or let them make this decision.

Comment (shalvah) [reply src/HoneybadgerServiceProvider.php:192]: They can do that easily by appending `|| true`.

As it is, what happens here would lead to a user's config being wrong _and they'd never know_, because how many times do you manually check your CI output? All you want to know is that it passes or fails, and then you investigate.

> We're not hiding the feedback in the console - the message is printed in the console.

No, we are...you only print the message, but not the stack trace, which shows where it originated from. And that's super important for _unexpected_ situations. (You have a `catch (Exception)` which means things you don't expect will be caught.)

File: src/Commands/HoneybadgerCheckinsSyncCommand.php
Comment (shalvah) [reply src/Commands/HoneybadgerCheckinsSyncCommand.php:49]: > I will catch ServiceException and log a warning message

This is a good approach. But that's not a warning. If my quota is up and my checkins aren't syncing, that's bad.

My take is that you should catch expected exceptions (`ServiceException`), log an error and fail (probably something like `$this->fail()`). Other errors should also fall automatically (but we should have good testing to limit unexpected crashes).

A user can always bypass this by adding `|| true` in their script (command returns a nonzero exit code, `|| true` makes it always return true).

**Comment by shalvah** [comment]
> Maybe `honeybadger:sync-checkins`

Yes, the prefix was implied. :) That works. Either that or `honeybadger:checkins:sync` (domain:action).

File: src/HoneybadgerServiceProvider.php
Comment (shalvah) [reply src/HoneybadgerServiceProvider.php:192]: I mean, it may lead to some annoyances or unexpected situations, but the downside is that:
- the user's checkins won't work and they won't know
- the package will be broken and we won't know

For runtime use, we can avoid this; for deployments, I think we can take the risk. We just need to have a high level of quality to ensure there are very few unexpected errors (like errors in code).

File: src/Commands/HoneybadgerCheckinsSyncCommand.php
Comment (subzero10) [reply src/Commands/HoneybadgerCheckinsSyncCommand.php]: There's no `$this->fail()` for commands, so I just let them all fail, which prints a helpful error message (when it's a `ServiceException`).

File: src/HoneybadgerServiceProvider.php
Comment (subzero10) [reply src/HoneybadgerServiceProvider.php]: Done!

**Comment by subzero10** [comment]
Hey @shalvah and @BethanyBerkowitz, I reworked the code to throw and fail the process in case of errors.
Thank you for all your feedback!

**Pull request merged**
